Panel: New - Favorite Widgets Section [ED-1501]

etayf opened the pull request.

Commit:
Favorite Widgets Section: New - add new section Favorites on top of panel (ED1501)

Commit:
Merge remote-tracking branch 'elementor/develop' into favorites-widgets-section

Commit:
Merge remote-tracking branch 'elementor/release/3.1.0' into favorites-widgets-section

Commit:
Favorite Widgets Section: Changed - add condition to add.js route (ED1501)

Commit:
Favorite Widgets Section: Changed - add static functions (ED1501)

KingYes committed:
Merge pull request #13655 from Nevoss/ED-1657

Compatibility-Tag: Fix - UI fixes for the message.

Commit:
Favorite Widgets Section: Changed - fixed routes and add.js (ED1501)

Commit:
Favorite Widgets Section: Changed - fixed routes and add.js (ED1501)

Commit:
Merge remote-tracking branch 'elementor/release/3.1.0' into favorites-widgets-section

Commit:
Favorite Widgets Section: Changed - fixed routes and add.js (ED1501)

Commit:
Merge remote-tracking branch 'etayf/favorites-widgets-section' into favorites-widgets-section

Commit:
Favorite Widgets Section: Changed - fixed add.js (ED1501)

Commit:
Favorite Widgets Section: Changed - try to add "inner-section" to widgetType ( waiting to answer ) (ED1501)

Commit:
Favorite Widgets Section: Changed -  ( waiting to answer ) (ED1501)

Commit:
Favorite Widgets Section: Changed -  waiting to answer section (ED-1501)

Commit:
Favorite Widgets Section: Changed -  unit test not completed (ED-1501)

Commit:
Favorite Widgets Section: Add - docs & disable inner-section  (ED-1501)

add function to disable right click ("Add to favorites") for section widget. it's temporary solution for visual bugs.

Commit:
Favorite Widgets Section: Add - disable inner-section  (ED-1501)

 add function to disable right click ("Add to favorites") for section widget. it is temporary solution for visual bugs.

Commit:
Favorite Widgets Section: Add - Docs  (ED-1501)

Commit:
Favorite Widgets Section: Change - remove docs from task  (ED-1501)

The docs move to task ED-1820

Commit:
Merge remote-tracking branch 'etayf/favorites-widgets-section' into favorites-widgets-section

Commit:
Favorite Widgets Section: Change - remove docs from task  (ED-1501)

The docs move to task ED-1820

Commit:
Favorite Widgets Section: Change - add helper for ignore widgets  (ED-1501)

ready for review without unittest

matipojo commented on modules/widgets-panel/assets/js/editor/commands/remove.js:
```diff
@@ -0,0 +1,27 @@
+		const result = await $e.data.delete( `panel-favorites/favorites?id=${ widget }`, {} );
+		if ( result.data ) {
+			// Get the categories of widget from front-end
+			const widgetsArr = elementor.widgetsCache[ widget ].categories;
```
Fix the name to fit his content. this variable does not contain a widget.

matipojo commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	addToolTip() {
+		const favoriteCategory = jQuery( '#elementor-panel-category-favorites .elementor-panel-category-title > i' );
```
jQuery elements should be prefixed with a `$`.

matipojo commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,140 @@
+
+	addToolTip() {
+		const favoriteCategory = jQuery( '#elementor-panel-category-favorites .elementor-panel-category-title > i' );
+		if ( favoriteCategory ) {
```
It's always true because jQuery always returns an object. use  `$favoriteCategory.length`.

matipojo commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,140 @@
+		// Remove the menu if exists
+		this.hideFavoriteWidgetsMenu();
+
+		const panel = jQuery( '#elementor-panel' );
```
$

etayf replied:
Fixed

matipojo commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,140 @@
+		// Find the category where user clicked
+		this.clickedWidgetCat = this.clickedWidget.parents( '.elementor-panel-category' ).attr( 'id' );
+		// Decided witch msg display. if clicked on widget under "Favorites" category display: Remove... otherwise Add...
+		const menuText = 'elementor-panel-category-favorites' === this.clickedWidgetCat ? 'Remove From Favorites' : 'Add To Favorites';
+		// Convert action text to class name
+		this.menuClass = menuText.toLowerCase().replaceAll( ' ', '-' );
+
+		// Create menu element with content
+		this.favoriteWidgetMenu = jQuery( '<div>', { id: this.menuId, class: this.menuClass } );
+		this.favoriteWidgetMenu.html( jQuery( '<span>' ).text( menuText ) );
```
Create different elements for each menu, so that on click you don't need to guess what the user wants.

etayf replied:
Fixed

matipojo commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,140 @@
+			if ( 'panel/elements/categories' === route ) {
+				this.menuId = 'e-favorite-widget-context-menu';
+				this.favoriteWidgetMenu = null;
+				this.clickedWidget = null;
```
Why do you need this property?

etayf replied:
Fixed

matipojo commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	addFavoriteWidgetsEvents() {
+		jQuery( document ).bind( 'contextmenu click keyup', ( e ) => {
```
Don't listen to the whole document.

etayf replied:
Fixed

matipojo commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,140 @@
+		}
+	}
+
+	addFavoriteWidgetsMenu( e ) {
```
Convert to a command.

etayf replied:
Fixed

matipojo reviewed (changes requested):
It's a very initial review.

Commit:
Favorite Widgets Section: Change - helper for ignore widgets  (ED-1501)

Commit:
Favorite Widgets Section: Fixed - logic and functions  (ED-1501)

Waiting to answer : what is the convenient way to move "favorites" category to top of categories (panel view)

etayf commented on modules/widgets-panel/assets/js/editor/commands/add.js:
```diff
@@ -0,0 +1,26 @@
+		const result = await $e.data.create( `panel-favorites/favorites?id=${ widget }`, {} );
+		if ( result.data ) {
+			// Get the categories of widget from front-end
+			const widgetsArr = elementor.widgetsCache[ widget ].categories;
```
Fixed

etayf commented on modules/widgets-panel/assets/js/editor/commands/add.js:
```diff
@@ -0,0 +1,26 @@
+
+export class Add extends CommandBase {
+	async apply( args = {} ) {
+		const widget = args.widget ?? '';
+
+		if ( ! widget.length ) {
+			return false;
+		}
+
```
Not fixed

matipojo replied:
Why?

etayf replied:
After we talk - Fixed

etayf commented on includes/editor-templates/panel-elements.php:
```diff
@@ -36,7 +38,7 @@
 
 <script type="text/template" id="tmpl-elementor-element-library-element">
-	<div class="elementor-element">
+	<div class="elementor-element" data-id="{{ obj.widgetType || obj.title.trim().toLowerCase().replaceAll( ' ', '-' ) }}">
```
Fixed

etayf commented on includes/editor-templates/panel-elements.php:
```diff
@@ -25,7 +25,9 @@
 <script type="text/template" id="tmpl-elementor-panel-elements-category">
-	<div class="elementor-panel-category-title">{{{ title }}}</div>
+	<div class="elementor-panel-category-title">{{{ title }}}
+		<# if ( 'favorites' === title.trim().toLowerCase() ) { #><i class="eicon-info-circle" data-tooltip="Right click on widgets to add or remove them from favorites"></i><# } #>
```
Fixed

etayf commented on modules/widgets-panel/assets/js/editor/commands/add.js:
```diff
@@ -0,0 +1,26 @@
+			// Get element position in array
+			const elmPos = widgetsArr.indexOf( 'favorites' );
+			// Check - if element in array do nothing, else add it to array
+			if ( -1 === elmPos ) {
```
Fixed

matipojo commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,171 @@
+	}
+
+	removeFavoriteMenu() {
+		const $favoriteWidgetMenu = jQuery( `#${ this.$favoriteMenuId }` );
```
Use the saved element.

etayf replied:
We talk about it - Fixed

matipojo commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,171 @@
+
+	removeFavoriteMenu() {
+		const $favoriteWidgetMenu = jQuery( `#${ this.$favoriteMenuId }` );
+		if ( $favoriteWidgetMenu.length ) {
```
The `remove` check also if exist.

etayf replied:
Fixed

matipojo commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,171 @@
+
+	contextmenuAddToFavorites() {
+		const $panelCategories = jQuery( '[ id^=elementor-panel-category- ]:not( [ id=elementor-panel-category-favorites ] )' );
+		if ( $panelCategories.length ) {
```
How can it be empty?

etayf replied:
Fixed

matipojo commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,171 @@
+		}
+	}
+
+	contextmenuRemoveFromFavorites() {
```
Please review the rest of the code according to the comments until here.

etayf replied:
We talk about it - Fixed

etayf commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,171 @@
+		$e.routes.on( 'run:after', ( component, route ) => {
+			if ( 'panel/elements/categories' === route ) {
+				this.$panel = jQuery( '#elementor-panel' );
+				this.$favoriteMenuId = 'e-favorite-widget-context-menu';
```
Fixed

etayf commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,171 @@
+			exit: {
+				keys: 'esc',
+				dependency: () => {
+					return jQuery( `#${ this.$favoriteMenuId }` ).length;
```
Fixed

etayf commented on modules/widgets-panel/assets/js/editor/commands/exit.js:
```diff
@@ -0,0 +1,9 @@
+
+export class Exit extends CommandBase {
+	apply() {
+		jQuery( `#${ this.component.$favoriteMenuId }` ).remove();
```
Fixed

etayf commented on modules/widgets-panel/assets/js/editor/commands/add.js:
```diff
@@ -0,0 +1,31 @@
+			// Get the categories of widget from front-end
+			const widgets = elementor.widgetsCache[ `${ widget }` ];
+			if ( widgets ) {
+				if ( ! widgets.hasOwnProperty( 'categories' ) ) {
+					widgets.categories = [];
+				}
```
Fixed

etayf commented on modules/widgets-panel/assets/js/editor/commands/add.js:
```diff
@@ -0,0 +1,31 @@
+		const result = await $e.data.create( `panel-favorites/favorites?id=${ widget }`, {} );
+		if ( result.data ) {
+			// Get the categories of widget from front-end
+			const widgets = elementor.widgetsCache[ `${ widget }` ];
```
Fixed

etayf commented on modules/widgets-panel/assets/js/editor/commands/add.js:
```diff
@@ -0,0 +1,31 @@
+					widgets.categories = [];
+				}
+				// Get element position in array
+				const elementExists = widgets.categories.includes( 'favorites' );
```
Fiexd

etayf commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,171 @@
+	setPositionMenu( favoriteWidgetMenu, e ) {
+		const menuWidth = favoriteWidgetMenu.outerWidth();
+		const setMenuPositionX = elementorCommon.config.isRTL ? ( e.pageX - menuWidth ) : e.pageX;
+		if ( favoriteWidgetMenu.length ) {
```
Fixed

etayf commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,171 @@
+		// Remove the menu if exists
+		this.removeFavoriteMenu();
+
+		const menuClass = menuText.toLowerCase().replaceAll( ' ', '-' );
```
Fixed

etayf commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,171 @@
+	addToolTip() {
+		const $favoriteCategory = jQuery( '#elementor-panel-category-favorites .elementor-panel-category-title' );
+		if ( 'favorites' === $favoriteCategory.text().trim().toLowerCase() ) {
+			const $iconInfo = jQuery( '<i>', { class: 'eicon-info-circle', 'data-tooltip': __( 'Right click on widgets to add or remove them from favorites', 'elementor' ) } );
```
Fixed

etayf commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,171 @@
+
+	addToolTip() {
+		const $favoriteCategory = jQuery( '#elementor-panel-category-favorites .elementor-panel-category-title' );
+		if ( 'favorites' === $favoriteCategory.text().trim().toLowerCase() ) {
```
Fixed

Commit:
Favorite Widgets Section: Fixed - review logic and functions  (ED-1501)

Commit:
Favorite Widgets Section: Fixed - review logic and functions  (ED-1501)

etayf commented on modules/widgets-panel/assets/js/editor/favorites-component.js:
```diff
@@ -0,0 +1,171 @@
+			$iconInfo.appendTo( $favoriteCategory );
+			const icon = $favoriteCategory.find( '> i' );
+			icon.tipsy( {
+				title: function title() {
```
We talk about it - the tipsy work like this.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/elementor/elementor?pullRequest=13635) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/elementor/elementor?pullRequest=13635) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: KingYes<br/>:x: Etay Frank<br/><hr/>**Etay Frank** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/elementor/elementor?pullRequest=13635) it.</sub>
